File: backend/app/services/reports 2.py
"""Validated test report listing."""
from app.models.order import Order, OrderTest
from app.schemas.enums import TestStatus
from app.schemas.order import OrderResponse, OrderTestResponse
from app.schemas.reports import ValidatedTestReportItem
from sqlalchemy.orm import Session, joinedload


class ReportService:
    def __init__(self, db: Session):
        self.db = db

    def list_validated_tests(self, limit: int = 500) -> list[ValidatedTestReportItem]:
        order_tests = (
            self.db.query(OrderTest)
            .options(
                joinedload(OrderTest.test),
                joinedload(OrderTest.order).joinedload(Order.patient),
                joinedload(OrderTest.order).selectinload(Order.tests).joinedload(OrderTest.test),
            )
            .filter(OrderTest.status == TestStatus.VALIDATED)
            .order_by(OrderTest.resultValidatedAt.desc())
            .limit(limit)
            .all()
        )

        items: list[ValidatedTestReportItem] = []
        for order_test in order_tests:
            order = order_test.order
            patient = order.patient
            order_payload = OrderResponse.model_validate(order).model_dump(mode="json")
            test_payload = OrderTestResponse.model_validate(order_test).model_dump(mode="json")
            items.append(
                ValidatedTestReportItem(
                    testId=order_test.id,
                    testCode=order_test.testCode,
                    testName=order_test.test.name if order_test.test else order_test.testCode,
                    orderId=order.orderId,
                    orderDate=order.orderDate,
                    patientId=patient.id,
                    patientName=patient.fullName,
                    patientDob=patient.dateOfBirth,
                    patientGender=patient.gender,
                    test=test_payload,
                    order=order_payload,
                )
            )
        return items

File: frontend/src/features/reports/api/reports 2.ts
/**
 * Report listing API — validated tests from server projection.
 */
import { apiClient } from '@/lib/api/client';
import { WORKFLOW_QUERY_LIMIT } from '@/lib/api/constants';
import type { Order, OrderTest } from '@/types';
import { useQuery } from '@tanstack/react-query';
import { useAuthStore } from '@/app/authStore';
import { queryKeys, cacheConfig } from '@/lib/query';
import type { ValidatedTest } from '../types';
import { calculateAge } from '@/utils';

export interface ValidatedTestReportItem {
  testId: number;
  testCode: string;
  testName: string;
  orderId: number;
  orderDate: string;
  patientId: number;
  patientName: string;
  patientDob?: string | null;
  patientGender?: string | null;
  test: OrderTest;
  order: Order;
}

function toValidatedTest(item: ValidatedTestReportItem): ValidatedTest {
  const age = item.patientDob ? calculateAge(item.patientDob) : undefined;
  return {
    testId: item.testId,
    testCode: item.testCode,
    testName: item.testName,
    orderId: item.orderId,
    orderDate: item.orderDate,
    patientId: item.patientId,
    patientName: item.patientName,
    patientAge: age,
    patientGender: item.patientGender ?? undefined,
    test: item.test,
    order: item.order,
  };
}

export const reportsAPI = {
  getValidatedTests(limit = WORKFLOW_QUERY_LIMIT): Promise<ValidatedTestReportItem[]> {
    return apiClient.get<ValidatedTestReportItem[]>('/reports/validated-tests', {
      limit: String(limit),
    });
  },
};

export function useValidatedTestsReport() {
  const { isAuthenticated, isLoading: isRestoring } = useAuthStore();
  const query = useQuery({
    queryKey: queryKeys.reports.validatedTests(),
    queryFn: () => reportsAPI.getValidatedTests(),
    enabled: isAuthenticated && !isRestoring,
    ...cacheConfig.dynamic,
    select: (items: ValidatedTestReportItem[]) => items.map(toValidatedTest),
  });

  return {
    validatedTests: query.data ?? [],
    isLoading: query.isLoading,
    isError: query.isError,
    error: query.error,
    refetch: query.refetch,
  };
}
